proposals/relaxed-simd/entropy.md

```diff
@@ -0,0 +1,57 @@
+# Entropy, Compat and Usage patterns
+
+The Relaxed SIMD proposal adds new operations that take advantage of the underlying hardware for accelerated performance by loosening the portability constraints of fixed-width SIMD, or adding new instructions that introduce local non-determinisim. This introduces a few potential risks specifically for browser engine implementations, though it’s possible that one or more of these might be applicable to other environments: 
+* Identifying underlying characteristics of the device (processor information if the engine that implements this proposal used the most optimal lowerings)
+* Possibly identifying information about the browser currently in use (i.e. if a sufficiently motivated user writes a hand tuned Wasm function when engines use different lowerings for instructions)
+* Compatibility risks, i.e. if code compiled for one browser works differently on a different browser (This is specific to observable behavior changes, and not performance differences)
+
+This document is an attempt to collate information that is already available in the issues filed for each of the operations on fingerprinting risk, quantify the risk for fingerprinting/compat issues and provide some information about usage patterns. 
+
+## Instruction summary
+
+* **relaxed i8x16.swizzle**<br>
+  *Entropy exposed:* Differences between Intel/ARM/RISC-V<br>
+  *Deterministic lowering:* Available<br>
+  *Compat risk:* Low, as the differences exposed are for out of range indices<br>
+  [*Issue link*](https://github.com/WebAssembly/relaxed-simd/issues/21)
+* **relaxed i32x4.trunc_{f32x4, f64x2} operations**<br>
+  *Entropy exposed:* Differences between Intel/ARM/RISC-V<br>
+  *Deterministic lowering:* Available<br>
+  *Compat risk:* Low, as the differing behavior is for out of range values, and NaNs<br>
+  [*Issue link*](https://github.com/WebAssembly/relaxed-simd/issues/21)<br>
+* **qfma, qfmns**<br>
+  *Entropy exposed:* Differences between hardware that has native FMA support, and hardware that does not.<br>
+  *Deterministic lowering:* Not available, but depending on underlying hardware, the results can only be fused, or unfused.<br>
+  *Compat risk:* TBD<br>
```

Comment:
My guess here is that when a Wasm binary that is used across different browsers that uses the fused operations, at the application level the only visible difference would be performance, so the compat risk is low. From early conversations about this, my understanding is that algorithms don't rely on the results of these operations being fused or not. @Maratyszcza does this sound right?

Reply:
The detection of fused/unfused results is captured in the entropy that this instruction exposes. To assess the compat risk, the information I'm looking for specifically is what is visible to an end application when a Wasm binary is run on a device that natively has FMA support vs. one that doesn't. With the way these instructions are currently written up, my expectation was that with the same wasm binary, the different codegen generated for the same Wasm binary will not be explicitly visible to the end application/benchmark that is using the Wasm binary. I'm also willing to try this out by compiling and running one of these benchmarks by forcing different codegen if you're able to link something here that we expect to be ported to Wasm, or is already in use.

Reply:
It was @Maratyszcza that mentioned this in the latest meeting.

One example requiring FMA I can find is multiplication by constant broken up into two values: `c1 * x + c2 * x`, `c1 + c2` representing a constant that is one step higher in precision. FMA provides one less rounding, which will affect the result. Example of the contrary is `(a*b) - (a*b)`, it can be lowered to to an FMA consisting of `a*b` and a negated result of another FMA also calculating `a*b`, the result being rounding error instead of zero. These differences in output are not guaranteed to be fatal immediately, but iterative algorithms can accumulate errors leading to divergence.

@Maratyszcza what did you have in mind?

Reply:
The main concern is error-free multiplication primitive. There are two ways to implement it: one using Dekker splitting and another using **true** FMA. Unfortunately, Dekker splitting gets completely broken when it is evaluated without intermediate rounding. However, people who use error-free multiplication primitive are aware of this, and I expect that software implementing this method, should it get ported to WAsm Relaxed SIMD, would detect fused/non-fused implementation of relaxed FMA and choose the appropriate code path. The software already has to do it as some architectures don't support FMA (e.g. current WebAssembly) and some always support it (e.g. ARM64).

Reply:
Right: Re: Veltkamp-Dekker.

We expect the developers who care would test the `qfma` operation itself. No plans to provide detection as a part of the proposal.